### PR TITLE
Feat/public letter route

### PR DESCRIPTION
**Pull request by gweiying**
## Context
We want to change our public letters route from `letters.gov.sg/letters/<id>` to `letters.gov.sg/<id>`, so that the public will be able to recognize our links more easily.

## Approach

- Routing
  - Set up frontend route for public letters at `letters.gov.sg/<id>`
  - Set up redirect to forward old route `letters.gov.sg/letters/<id>` to new route `letters.gov.sg/<id>`
  - Remove old catchall route redirects (`letters.gov.sg/*` used to redirect to `letters.gov.sg/error`)
- Update frontend links/ buttons to the letters
- Added logic to defend protected namespaces during public id generation

## Screenshots
Screenshot 2023-06-23 at 5 23 35 PM

Redirection:
gif

### CODE REVIEW

File: backend/src/core/utils/generate-public-id.spec.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    // Replace the original generateId function with the mock function
+    const id = generatePublicId(mockGenerateId)
+
+    // Expect the mock generateId function to be called twice (once for the initial match and once for the regenerated ID)
+    expect(mockGenerateId).toHaveBeenCalledTimes(2)
+    // Expect the regenerated ID to be different from the protected string
+    expect(PROTECTED_NAMESPACES).not.toContain(id)
+    // Expect the output to be 'abcdefghi'
+    expect(id).toEqual('abcdefghi')
```
Comment (AikoPath) [inline backend/src/core/utils/generate-public-id.spec.ts:26]: I am not sure if the comments here that explain what the code does are required.

File: frontend/src/app/routes/public.routes.tsx [hunk 2]
```diff
@@ -31,8 +37,4 @@ export const publicRoutes: RouteObject[] = [
       },
     ],
   },
-  {
-    path: '*',
-    element: <Navigate to={routes.public.error} />, // TODO: Differentiate between invalid letters link and invalid page?
-  },
 ]
```
Comment (AikoPath) [inline frontend/src/app/routes/public.routes.tsx:40]: This PR breaks the /error re-direct for paths with more than one path component (other than `/admin/`)
I suspect this is the part of the PR that causes this.

Screenshot 2023-06-23 at 5 32 05 PM
Screenshot 2023-06-23 at 5 27 54 PM

**Review by AikoPath — CHANGES_REQUESTED** [review]
Looks like this is breaking the existing 404 error handling handling.

File: frontend/src/app/routes/public.routes.tsx (hunk 2 above)
Comment (gweiying) [reply frontend/src/app/routes/public.routes.tsx:40]: Good catch will look into it, thanks!

Comment (gweiying) [reply frontend/src/app/routes/public.routes.tsx]: Reverted in [b49f919](https://github.com/opengovsg/letters/pull/161/commits/b49f919e747c98a045a0fe00a003ad384dd81df8) and updated in [7ec58b7](https://github.com/opengovsg/letters/pull/161/commits/7ec58b77cc1252e059177bd23f24b83c89825098).

The change now causes the following behavior:
- if a person lands on [letters.gov.sg/](http://letters.gov.sg/)<invalid-letters-link>, we redirect them to [letters.gov.sg/error](http://letters.gov.sg/error)
- if a person lands on [letters.gov.sg/hello/there](http://letters.gov.sg/hello/there) (invalid link), we redirect them to [letters.gov.sg/error](http://letters.gov.sg/error)
- if a person lands on [letters.gov.sg/admin/hello](http://letters.gov.sg/admin/hello) (internal link), we redirect them to [letters.gov.sg/admin/login](http://letters.gov.sg/admin/login) if they're not logged in.

File: backend/src/core/utils/generate-public-id.spec.ts (hunk 1 above)
Comment (gweiying) [reply backend/src/core/utils/generate-public-id.spec.ts]: [d9d167c](https://github.com/opengovsg/letters/pull/161/commits/d9d167cacf7e95883c86580050a7102b1911a4a7)

**Review by AikoPath — APPROVED** [review]
LGTM

**Pull request merged**